util/argo/resource_tracking.go

```diff
@@ -155,6 +156,13 @@ func (rt *resourceTracking) SetAppInstance(un *unstructured.Unstructured, key, v
 		}
 		if len(val) > LabelMaxLength {
 			val = val[:LabelMaxLength]
+			// Prevent errors if the truncated name ends in a special character.
+			// See https://github.com/argoproj/argo-cd/issues/18237.
+			okEndPattern := regexp.MustCompile("[a-zA-Z0-9]$")
+
+			for !okEndPattern.MatchString(val) {
```

Comment:
This is paranoia realm, but I think it'd be wise to also check for `len(val) > 1` here, just in case someone will ever be able to pass a `val` that would trigger an out of bound exception.